### PR TITLE
Collection Comprehension

### PR DESCRIPTION
**Pull request by yloiseau**
Several constructions are allowed:

```golo
  list[ x * 2 for x in aList ]
  set[ a + b for a, b in aTupleList]
  [ a + b for a in [1..6] for b in [1..6] ]
  map[ [k, v] for k, v in aMap: entrySet() ]
```

This is not finished yet, needs more polishing (and doc...)
Comments ?

### CODE REVIEW

**Comment by jponge** [comment]
Lovely, really!

**Comment by yloiseau** [comment]
:smile:

**Comment by rlespinasse** [comment]
Nice

**Comment by jponge** [comment]
How far are we from a full code review?

**Comment by yloiseau** [comment]
I think I'm OK with the code, but I didn't had much review feedback.

I'd like to add another syntax `iter[comprehension expression]` that generate a lazy iterator, but I wanted to add it in another PR. However, it can be added in this one I you prefer. My goal is to create a module and augmentations to deal with lazyness (predefined generators, HOF, and so on), using this generator class and lazy lists.

**Comment by jponge** [comment]
Just let me know when you want a code review then, I’ll do it.

**Comment by yloiseau** [comment]
I think it's ready for a review ;)

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java [hunk 1]
```diff
@@ -35,6 +36,10 @@
 
   private GoloCompilationException.Builder exceptionBuilder;
 
+  private static String syntheticName(String base) {
+    return "__$$_" + base + "_" + System.currentTimeMillis();
```
Comment (Artpej) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: I'm not sure what exactly append here :wink:, but it seems to me possible to generate the same synthetic name both if we call this function two times in the same millisecond. Why not use an increment based on `Context` like with closure and decorators synthetic names?

Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: Indeed. I actually only extracted existing code. BTW, a more generic implementation will be introduced by the macro branch (mainly for hygienic aspect but reusable in syntactic transformations like these)

File: doc/basics.asciidoc [hunk 2]
```diff
@@ -409,7 +409,153 @@ let b = -> [2, 'b']
...
+----
+
+This is a more readable and more powerful version of `filter`+`map`. The
+previous example could be rewritten as
+[source,golo]
+----
+let l2 = l2: filter(|x| -> (x % 2) == 0): map(|x| -> x * 2)
+----
+
+The general syntax is an collection literal containing an expression followed
```
Comment (jponge) [inline doc/basics.asciidoc]: is _a_ collection literal

File: doc/basics.asciidoc [hunk 3]
```diff
@@ -409,7 +409,153 @@ let b = -> [2, 'b']
...
+Contrary to the `filter`+`map` approach, where the kind on collection is kept,
+comprehension can transform the source collection type, which can be any
+iterable. For instance:
+[source,golo]
+----
+let dices = set[ x + y foreach x in [1..7] foreach y in [1..7]]
+# dices == set[2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
+----
+
+However, the result collection can only be one of the predefined collection
```
Comment (jponge) [inline doc/basics.asciidoc]: Perhaps say it's of the type of one of the collection literal tyoes

File: doc/basics.asciidoc [hunk 4]
```diff
@@ -409,7 +409,153 @@ let b = -> [2, 'b']
...
+
+xref:_destructuring[Destructuring] can also be used in collection
+comprehension, as in
+[source,golo]
+----
+let couples = [ [1, 2], [2, 3], [3, 4] ]
+let sums = [ a + b foreach a, b in couples ]
+----
+
+Maps can also be created, provided the given expression is either a 2-tuple or
```
Comment (jponge) [inline doc/basics.asciidoc]: A _pair tuple_ rather than a _2-tuple_?

File: doc/basics.asciidoc [hunk 5]
```diff
@@ -409,7 +409,153 @@ let b = -> [2, 'b']
...
+----
+foreach v in [[x,x] foreach x in [0..3]] {
+  println(v)
+}
+----
+
+==== Collection comprehension vs. `map` and `filter`
+
+Collection comprehension is actually quite similar to using `map` and `filter`
+HOF on a collection. Indeed, a comprehension such as:
```
Comment (jponge) [inline doc/basics.asciidoc]: _HOF_ `->` _higher-order function_

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java (hunk 1 above)
Comment (jponge) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: I'm the one to blame, as Yannick said I've done that in other places for synthetic variables.

Perhaps we should just eliminate all these time-based generations in the whole code, using simple atomic long integers. Thoughts?

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java [hunk 6]
```diff
@@ -463,11 +468,66 @@ public Object visit(ASTCollectionLiteral node, Object data) {
...
+      context.objectStack.push(comprehension);
+      return data;
+    }
     CollectionLiteral.Type type = CollectionLiteral.Type.valueOf(node.getType());
     context.objectStack.push(new CollectionLiteral(type, expressions));
     return data;
   }
 
+  private Block createCollectionComprehension(Context context, String type, List<ExpressionStatement> expressions) {
+    // TODO: special case for iter that create a generator
```
Comment (jponge) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: Still a todo?

File: src/test/resources/for-execution/comprehension.golo [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    [2, 2]
+  ]
+  let l1 = list[ [x, y] foreach x in [0..3] for (var y = 0, y < 3, y = y + 1) ]
+  require(l1 == ref, "two_mixed_loops failed")
+  let l2 = list[ [x, y] for (var x = 0, x < 3, x = x + 1) foreach y in [0..3] ]
+  require(l2 == ref, "two_mixed_loops failed")
+}
+
+function test_more_loops = {
+  let l = list[ [x,y,z] foreach x in [0,1,2] when (x % 2) == 1
```
Comment (jponge) [inline src/test/resources/for-execution/comprehension.golo]: :heart:

**Comment by jponge** [comment]
See my minor comments.

This is a very beautiful feature!

Maybe you should ping folks on `golo-dev` for further comments.

**Comment by jponge** [comment]
BTW I would target this for `3.0.0-incubation-M2`, not the first milestone that we will have once the Hudson builds work.

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java (hunk 6 above)
Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: yes, this feature will be added in a separate PR dealing more specifically with generators and lazyness. Do you rather I remove the TODO comment (or make it a NOTE)?

**Comment by yloiseau** [comment]
@jponge you're the boss on release planning matters :smile:

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java (hunk 6 above)
Comment (jponge) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: No you can leave it, I was just wondering :-)

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java (hunk 1 above)
Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: As I said, I already made a alternative implementation in the macro branch (see https://github.com/yloiseau/golo-lang/blob/wip/macro-quote/src/main/java/gololang/macros/SymbolGenerator.java). However, since I don't know when it will be ready, I can already make change (simply unsing a counter) in this PR.

Comment (jponge) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: :+1:

Comment (Artpej) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: *"However, since I don't know when it will be ready, I can already make change (simply using a counter) in this PR."*
:+1:

**Comment by jponge** [comment]
Good to rebase and merge at your convenance @yloiseau
